cpu manager: specify the container CPU set during the creation

cynepco3hahue opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:
We can set the container `cpuset.cpus` during the creation and it
will not need to call to update resources after the container creation.

The additional side effect of the change, that the runc process that is responsible
to create the container will run with the same CPU affinity because the
runc runs on the cpuset provided in the config.json arg.

It will prevent undesirable interrupts on isolated CPUs.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

Commit:
cpu manager: specify the container CPU set during the creation

We can set the container cpuset.cpus diring the creation and it
will not need to call to update resources after the container creation.

Additional side effect of the change, that the runc process that responsible
to create the container will run with the same CPU affinity because the
runc runs on the cpuset provided in the config.json arg.

It will allow to prevent undesirable interupts on isolated CPUs.

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

klueska commented:
I haven't looked closely at this yet, but can you elaborate on what bug this is fixing (i.e. you marked it as `/kind bug`) and what exactly you mean by:

> The additional side effect of the change, that the runc process that is responsible to create the container will run with the same CPU affinity because the runc runs on the cpuset provided in the config.json arg

Does the new hook you implement make sure that the list of allocated CPUs get passed to the runc config?
Does this work for all versions of runc?

klueska commented:
/assign klueska

cynepco3hahue commented:
> I haven't looked closely at this yet, but can you elaborate on what bug this is fixing (i.e. you marked it as `/kind bug`) and what exactly you mean by:
> 
> > The additional side effect of the change, that the runc process that is responsible to create the container will run with the same CPU affinity because the runc runs on the cpuset provided in the config.json arg

Hi Kevin, I wasn't totally sure about the kind, maybe setting it as cleanup will be more correct. But one thing for sure, once we are running some latency-sensitive workload under our environment it possible that some other container that crashed at the same time cause runc to be scheduled on the isolated CPU(caught via `ftrace`). We talked with one of the runc maintainers and he confirmed that the runc creation process runs on top of the CPU set that you passed to the container config during the creation. Currently, under the kubelet, we have the next flow creation, `cpumanager.Allocate -> create container -> update container resources(update cpuset.cpus) -> start container`, so in the current flow, the runc creation process runs on top of all available CPUs under the node because update the CPU set too late.

> 
> Does the new hook you implement make sure that the list of allocated CPUs get passed to the runc config?
> Does this work for all versions of runc?

Yes, it passes cpu set during the container creation so the CRI already gets the updated config. Regarding the second question, IIUC each Kubernetes version goes with the specific runc version, so in general, it should work only for runc that go together with the latest Kubernetes. It is hard to verify because the runc creation process exists a very short amount of time under the node.

dims commented:
/ok-to-test

klueska commented on pkg/kubelet/cm/internal_container_lifecycle.go:
```diff
@@ -37,13 +38,23 @@ type internalContainerLifecycleImpl struct {
 }
 
+func (i *internalContainerLifecycleImpl) PreCreateContainer(pod *v1.Pod, container *v1.Container, containerConfig *runtimeapi.ContainerConfig) error {
+	if containerConfig.Linux != nil && containerConfig.Linux.Resources != nil {
+		if i.cpuManager != nil {
+			allocatedCPUs := i.cpuManager.GetAllocatedResources(pod, container)
+			if allocatedCPUs != "" {
+				containerConfig.Linux.Resources.CpusetMems = allocatedCPUs
+			}
+		}
+	}
```
I don't think we simply want to check for `containerConfig.Linux != nil && containerConfig.Linux.Resources != nil` and then set CPUs in them if they exist. I think we want to explicitly create those structures if they do not exist and then add CPUs to them. Moreover, this should explicitly only be done on linux, so something like:
```
if i.cpuManager == nil {
    return nil
}

if runtime.GOOS != "linux" {
    return nil
}

allocatedCPUs := i.cpuManager.GetAllocatedResources(pod, container)
if allocatedCPUs == "" {
    return nil
}

if containerConfig.Linux == nil {
    containerConfig.Linux = ...
}

if containerConfig.Linux.Resources == nil {
    containerConfig.Linux.Resources = ...
}

containerConfig.Linux.Resources.CpusetMems = allocatedCPUs

return nil
```

cynepco3hahue replied:
I checked again the code and I can see that all needed structure initialized under https://github.com/kubernetes/kubernetes/blob/3b38f6a08e255b3eb2e34ffb3c3cecacc081f334/pkg/kubelet/kuberuntime/kuberuntime_container.go#L160 so I will just remove all redundant checks.

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

cynepco3hahue commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -63,7 +63,10 @@ type Manager interface {
+	AddContainer(p *v1.Pod, c *v1.Container, containerID string)
+
+	// GetAllocatedResources return the CPU affinity for the container
+	GetAllocatedResources(pod *v1.Pod, container *v1.Container) string
```
I just wanted to have an interface method with the neutral name because I will want to have the same for the memory manager, WDT?

klueska replied:
I still think we should just consolidate it down to a single call of:
```
GetCPUs(podUID, containerName string) cpuset.CPUSet
```

We should then expand `cpuset.CPUSet` to include helper functions for:
```
ToSliceInt64() []int64
ToSliceNoSortInt64() []int64
```

Then the two call sites that use it can each call functions directly attached to the `CPUSet` type to get the format that they want.

This should be added as a standalone commit before the rest of the changes here.

klueska replied:
I'm happy to rename the function to `GetAllocatedCPUs()` if you like, but I think it's better to keep it CPU specific.

cynepco3hahue replied:
Do you want a separate PR for it or the separate commit will be good enough for it?

klueska replied:
Separate commit is fine.

cynepco3hahue commented on pkg/kubelet/cm/internal_container_lifecycle.go:
```diff
@@ -37,13 +38,23 @@ type internalContainerLifecycleImpl struct {
 	topologyManager topologymanager.Manager
 }
 
+func (i *internalContainerLifecycleImpl) PreCreateContainer(pod *v1.Pod, container *v1.Container, containerConfig *runtimeapi.ContainerConfig) error {
```
Good catch! done

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

cynepco3hahue commented:
/retest

klueska commented on pkg/kubelet/cm/internal_container_lifecycle.go:
```diff
@@ -37,13 +38,22 @@ type internalContainerLifecycleImpl struct {
 	if i.cpuManager != nil {
-		err := i.cpuManager.AddContainer(pod, container, containerID)
-		if err != nil {
-			return err
+		allocatedCPUs := i.cpuManager.GetAllocatedResources(pod, container)
+		if allocatedCPUs != "" {
+			containerConfig.Linux.Resources.CpusetMems = allocatedCPUs
 		}
 	}
```
Does something else protect this from getting called on non.linux systems? Is creation of the CPUManager somehow only limited to linux?

klueska replied:
In fact, I see this in `pkg/kubelet/cm/container_manager_windows.go`
```
func (cm *containerManagerImpl) InternalContainerLifecycle() InternalContainerLifecycle {
    return &internalContainerLifecycleImpl{cpumanager.NewFakeManager(), topologymanager.NewFakeManager()}
}
```
Which will obviously cause problems on windows since I'm guessing it doesn't support `containerConfig.Linux.Resources.CpusetMems`.

cynepco3hahue replied:
We have two implementations of the container manager that compiled according to build flags, so if it windows specific build it will not include at all https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/internal_container_lifecycle.go

klueska replied:
I don't quite follow. The code I posted above clearly shows that the windows code instantiates an instance of `internalContainerLifecycleImpl`, passing it an instance of `cpumanager.NewFakeManager()`, meaning that `i.cpuManager != nil` will not be `nil` inside its call to `InternalContainerLifecycle()`.

cynepco3hahue replied:
Oh, you are right, but in this case, FakeManager will return an empty string during the call to `GetAllocatedResources`, and the `if` check will not pass.

klueska replied:
Yeah. So it looks like it wouldn't cause issues in its current form, but it's also very subtle why it doesn't.
I would still either:
1. Add a `runtime.GOOS == "linux"` check here; or
2. Split the implementation of this one function into separate `_linux.go`, `_windows.go` files.

cynepco3hahue replied:
I can see that we have the same thing for the `AddContainer` can we leave such refactoring to the separate PR or do you prefer to finish with it under the current one?

klueska replied:
`AddContainer()` is OK because it calls back into the `CPUManager `to do its thing. The real CPUManager will add the container to its internal containerMap and the fake CPUManager will do nothin (neither of which is linux specific). The only reason we need to branch for linux here is because we explicitly set a  linux specific field in `containerConfig.Linux.Resources.CpusetMems`.  If/when windows supports something similar, they can implement this differently using the same underlying CPUmanager.

cynepco3hahue replied:
Makes sense, will create different files for Linux and windows implementation.

Commit:
Provide additional methods under the CPUSet

- ToSliceInt64 returns sorted slice of cores IDs in int64 format
- ToSliceNoSortInt64 returns slice of cores IDs in int64 format

Signed-off-by: Artyom Lukianov <alukiano@redhat.com>

Commit:
Update CPU manager GetCPUs method to return pointer to CPUSet

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

cynepco3hahue commented:
@klueska I tested the PR locally and the reason why it fails under the CI, it because of the docker use. The fix works just fine with the container + runc.

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

cynepco3hahue commented:
@klueska CPU manager test passed can you please review again?

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -80,7 +80,7 @@ type Manager interface {
 	// GetCPUs implements the podresources.CPUsProvider interface to provide allocated
 	// cpus for the container
-	GetCPUs(podUID, containerName string) []int64
+	GetCPUs(podUID, containerName string) *cpuset.CPUSet
```
This should not be a pointer to a CPUSet, but just a CPUSet.
Under the hood it is just a struct wrapping a map, so in terms of storage, it is already just passing a pointer around.

cynepco3hahue replied:
I think the only overhead will be because of struct copy, why do you not like to use a pointer here?

klueska replied:
Mostly just because the `CPUSet` type was designed so that you wouldn't have to pass around a pointer to it.

It's already passed around efficiently as the CPUSet type itself. Passing a pointer to it is actually *less*  efficient, because the exact same size object is being passed around, but there is one more level of indirection you have to go through get to the underlying map.

In other words, the definiton of the `CPUSet` struct looks as follows:
```
type CPUSet struct {
    elems map[int]struct{}
}
```

Which (in terms of storage) is essentially the same as the following, since go stores `map`s as pointers:
```
type CPUSet struct {
    elems uintptr
}
```

Which then (in terms of storage) is exactly the same as a standalone `uintptr`.

So by passing a full `CPUSet` type, you are essentially just passing a direct copy of the underlying `uintptr` to its embedded map (with 0 extra overhead). By passing a pointer to a `CPUSet` you are still passing a `uintptr` around, but you have to follow one more level of indirection before getting to the underlying map's pointer.

klueska replied:
For reference:
https://play.golang.org/p/Eb0u32gTgDR

cynepco3hahue replied:
I see thanks for the explanation.

klueska replied:
For reference:
https://play.golang.org/p/WeCYE-5MY9O

klueska replied:
For reference:
https://play.golang.org/p/TlclpQYkBLS

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -63,7 +63,7 @@ type Manager interface {
 	// AddContainer is called between container create and container start
 	// so that initial CPU affinity settings can be written through to the
 	// container runtime before the first process begins to execute.
-	AddContainer(p *v1.Pod, c *v1.Container, containerID string) error
+	AddContainer(p *v1.Pod, c *v1.Container, containerID string)
```
The comment here is now incorrect.

cynepco3hahue replied:
Goot catch, done

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -481,11 +462,7 @@ func (m *manager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet)
-		result = append(result, int64(cpu))
-	}
-	return result
+func (m *manager) GetCPUs(podUID, containerName string) *cpuset.CPUSet {
+	cpus := m.state.GetCPUSetOrDefault(podUID, containerName)
+	return &cpus
```
Should not return a pointer type.

cynepco3hahue replied:
done

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -481,11 +462,7 @@ func (m *manager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet)
-func (m *manager) GetCPUs(podUID, containerName string) []int64 {
-	cpus := m.state.GetCPUSetOrDefault(string(podUID), containerName)
-	result := []int64{}
-	for _, cpu := range cpus.ToSliceNoSort() {
```
I see the old implementation of this used `ToSliceNoSort()`, but your updated use on https://github.com/kubernetes/kubernetes/pull/98019/files#diff-c07d343a8d459721656cf7ddacda71506bde01a5cac288d15a4aba6bfc9a1caeR1031 does not use a `NoSort()` variant.

cynepco3hahue replied:
nice catch, done

klueska commented on pkg/kubelet/cm/cpumanager/fake_cpu_manager.go:
```diff
@@ -69,7 +69,7 @@ func (m *fakeManager) State() state.Reader {
 }
 
-func (m *fakeManager) GetCPUs(podUID, containerName string) []int64 {
+func (m *fakeManager) GetCPUs(podUID, containerName string) *cpuset.CPUSet {
```
Should not be pointer

klueska commented on pkg/kubelet/cm/cpumanager/state/state_checkpoint_test.go:
```diff
@@ -227,8 +227,8 @@ func TestCheckpointStateRestore(t *testing.T) {
 				if strings.TrimSpace(tc.expectedError) != "" {
-					if strings.Contains(err.Error(), "could not restore state from checkpoint") &&
-						strings.Contains(err.Error(), tc.expectedError) {
+					tc.expectedError = "could not restore state from checkpoint: " + tc.expectedError
+					if strings.HasPrefix(err.Error(), tc.expectedError) {
```
This change seems unrelated and a bit sstrange to be updating `tc.expectedError` like this. What are you trying to achieve?

cynepco3hahue replied:
done

klueska commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -198,6 +199,29 @@ func (s CPUSet) ToSliceNoSort() []int {
+	for cpu := range s.elems {
+		result = append(result, int64(cpu))
+	}
+
```
nit: unnecessary newline

cynepco3hahue replied:
done

klueska commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -198,6 +199,29 @@ func (s CPUSet) ToSliceNoSort() []int {
+	for cpu := range s.elems {
+		result = append(result, int64(cpu))
+	}
+
```
nit: unnecessary newline

cynepco3hahue replied:
done

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

cynepco3hahue commented:
/retest

klueska commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -167,6 +167,13 @@ func (m *kubeGenericRuntimeManager) startContainer(podSandboxID string, podSandb
+	if err != nil {
+		s, _ := grpcstatus.FromError(err)
+		m.recordContainerEvent(pod, container, "", v1.EventTypeWarning, events.FailedToCreateContainer, "Internal PreCreateContainer hook failed: %v", s.Message())
+		return s.Message(), ErrPreStartHook
```
Is `ErrPreStartHook` the correct error type to return here? Should we add a new `ErrPreCreateHook` error?

cynepco3hahue replied:
True, done

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

klueska commented:
Thanks for putting this together @cynepco3hahue.

With the introduction of https://github.com/kubernetes/kubernetes/pull/87759, the logic to allocate exclusive CPUs to a pod was moved from pod creation time to pod admission time. As such, the full set of exclusive CPUs that should be allocated to a pod is now known before the pod itself ever gets created. This now gives us the opportunity to set each container's `cpuset` *before* it gets launched.

In the old way of doing things this was not the case. Since we didn't actually allocate exclusive CPUs to a container until *after* it was created, it was necessary to make an `Update()` call against the runtime API in order to update the `cpuset` it was created with.

Now that we know the CPUSet before container creation this is no longer necessary, and we can update it before the container actually starts. This PR adds the logic to do this in a clean and consistent manner.

/lgtm
/approve

cynepco3hahue commented:
@klueska A lot of thanks for the great review!

cynepco3hahue commented:
/retest

klueska commented:
Unfortunately the small changes to:
```
pkg/kubelet/dockershim/helpers_linux.go
pkg/kubelet/kuberuntime/kuberuntime_container.go
```
mean that this is no longer isolated to just `pkg/kubelet/cm` (where I have approval rights).

We will need to get someone with approval rights at the `pkg/kubelet` level to review as well.

/cc @derekwaynecarr

klueska commented:
/test pull-kubernetes-node-kubelet-serial-topology-manager

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

cynepco3hahue commented:
/retest

derekwaynecarr commented:
dockershim and kuberuntime changes all look good.

thanks for the thorough review @klueska 

/lgtm
/approve
